### PR TITLE
Add Playlist Functionality

### PR DESCRIPTION
**Pull request by candry7731**
Add Playlist functionality to Home/Grid screens. 

**Changes**
Added Playlist collection type

### CODE REVIEW

File: components/ItemGrid/GridItem.brs [hunk 1]
```diff
@@ -57,7 +57,11 @@ sub itemContentChanged()
     else if itemData.type = "Photo"
         m.itemPoster.uri = itemData.PosterUrl
         m.itemText.text = itemData.Title
-    else if itemData.type = "MusicArtist"
+    else if itemData.type = "Episode"
+        m.itemPoster.uri = itemData.PosterUrl
+        m.itemText.text = itemData.Title
+        print "Episode data: " itemData
```
Comment (1hitsong) [inline components/ItemGrid/GridItem.brs]: Debug code

File: components/ItemGrid/LoadItemsTask2.brs [hunk 2]
```diff
@@ -79,8 +79,10 @@ sub loadItems()
                 tmp = CreateObject("roSGNode", "FolderData")
             else if item.Type = "MusicArtist" or item.Type = "MusicAlbum"
                 tmp = CreateObject("roSGNode", "MusicArtistData")
-            else if item.Type = "Audio"
+            else if item.Type = "Audio" or item.Type = "AudioBook"
                 tmp = CreateObject("roSGNode", "MusicSongData")
+            else if item.Type = "Episode"
```
Comment (1hitsong) [inline components/ItemGrid/LoadItemsTask2.brs]: Adding this section causes a warning when you include a TV episode in your playlist:

`Warning occurred while setting a field of an RoSGNode
-- Tried to set nonexistent field "favorite" of a "TVEpisodeData" node
pkg:/components/ItemGrid/LoadItemsTask2.brs(93)`

**Review by 1hitsong — CHANGES_REQUESTED** [review]
Adding mixed media results in a lackluster viewing experience. Movies get artwork. Songs and TV shows don't.

As it stands, this change allows you to access and view playlists, and play each item one at a time. I think since this is for playlist functionality we need it to move from the item the user selects to the next.

I'd also like to see the items in the list view, similar to how songs are listed on album detail pages, and how the web view displays playlists. The grid layout doesn't indicate to the user that items will play in the order they appear.

WIN_20220702_08_37_32_Pro

**Comment by candry7731** [comment]
> Adding mixed media results in a lackluster viewing experience. Movies get artwork. Songs and TV shows don't.
> 
> As it stands, this change allows you to access and view playlists, and play each item one at a time. I think since this is for playlist functionality we need it to move from the item the user selects to the next.
> 
> I'd also like to see the items in the list view, similar to how songs are listed on album detail pages, and how the web view displays playlists. The grid layout doesn't indicate to the user that items will play in the order they appear.
> 
> WIN_20220702_08_37_32_Pro
> 

Can you check now? This has been updated to use the episode row scene

**Comment by 1hitsong** [comment]
@candry7731 I'm away from home until next week. I'll check it out when I return 🤘

**Comment by candry7731** [comment]
> @candry7731 I'm away from home until next week. I'll check it out when I return 🤘

Sounds good! Thanks

File: components/ItemGrid/GridItem.brs (hunk 1 above)
Comment (cewert) [reply components/ItemGrid/GridItem.brs]: @candry7731 this still needs addressed

File: components/ItemGrid/LoadItemsTask2.brs (hunk 2 above)
Comment (cewert) [reply components/ItemGrid/LoadItemsTask2.brs]: @candry7731 this still needs addressed

**Comment by cewert** [comment]
Marking this as draft till all the current issues are resolved. Please mark this PR as "ready for review" once everything is tested and you think it's ready to be merged

**Comment by 1hitsong** [comment]
This messes up the display of the Music library posters. Here's what I'm seeing for each branch.

Master
Master

Playlist
Playlist

**Comment by 1hitsong** [comment]
- [ ] On the main Playlist Grid View, let's use square posters instead of the tall ones.
- [ ] On the playlist list view, let's use square posters here as well, similar to the album detail screen
- [ ] On the playlist list view, let's make the rows smaller. I know it'll make the artwork difficult to really see, but in this view I think that's 100% fine.
- [ ] Movie posters in playlist list view are resized and squished to be square, making them look bad. Instead let's size it down and crop it to be square, like the webview does.
- [ ] Add Play button to playlist list view to make it play through the playlist from start to finish 😸 
- [ ] I played an episode from the playlist. At the end of the episode I expected it to play the next items in the playlist, instead it played the next episode in the series, which isn't in the playlist.

File: components/ItemGrid/ItemGrid.brs [hunk 3]
```diff
@@ -149,12 +149,15 @@ sub loadInitialItems()
         if get_user_setting("display.livetv.landing") = "guide" and m.options.view <> "livetv"
             showTvGuide()
         end if
-
-
+    else if m.top.parentItem.collectionType = "books"
```
Comment (1hitsong) [inline components/ItemGrid/ItemGrid.brs:197]: Dunno if we want/need to add any more book code at this moment.

File: components/music/NowPlaying.brs [hunk 4]
```diff
@@ -382,15 +382,27 @@ sub onMetaDataLoaded()
...
-        m.songDuration = data.json.RunTimeTicks / 10000000.0
+
+        if data.type = "AudioBook"
+            m.songDuration = data.RunTimeTicks / 10000000.0
+            setPosterImage(data.posterURL)
+            setScreenTitle(data)
+            setOnScreenTextValues(data)
+        else
+            m.songDuration = data.json.RunTimeTicks / 10000000.0
+            setPosterImage(data.posterURL)
```
Comment (1hitsong) [inline components/music/NowPlaying.brs]: This line is duplicated and can be moved outside of the conditional statement.

File: source/Main.brs [hunk 5]
```diff
@@ -111,7 +113,10 @@ sub Main (args as dynamic) as void
         else if isNodeEvent(msg, "selectedItem")
             ' If you select a library from ANYWHERE, follow this flow
             selectedItem = msg.getData()
-            if selectedItem.type = "CollectionFolder" or selectedItem.type = "UserView" or selectedItem.type = "Folder" or selectedItem.type = "Channel" or selectedItem.type = "Boxset"
+            if selectedItem.json.Type = "Playlist"
+                group = CreatePlaylistGroup(selectedItem.json)
+                'sceneManager.callFunc("pushScene", group)
```
Comment (1hitsong) [inline source/Main.brs:117]: Commented out code should be removed

**Review by 1hitsong — CHANGES_REQUESTED** [review]

File: components/ItemGrid/ItemGrid.brs (hunk 3 above)
Comment (candry7731) [reply components/ItemGrid/ItemGrid.brs:197]: This is to get the audiobook to display the parent folders when selecting audiobook form the home screen nav.

File: components/ItemGrid/LoadItemsTask2.brs (hunk 2 above)
Comment (candry7731) [reply components/ItemGrid/LoadItemsTask2.brs]: this has been fixed

**Comment by candry7731** [comment]
> * [X]  On the main Playlist Grid View, let's use square posters instead of the tall ones.
> * [X]  On the playlist list view, let's use square posters here as well, similar to the album detail screen
> * [X]  On the playlist list view, let's make the rows smaller. I know it'll make the artwork difficult to really see, but in this view I think that's 100% fine. 
> * [X]  Movie posters in playlist list view are resized and squished to be square, making them look bad. Instead let's size it down and crop it to be square, like the webview does. - **I set this to scaleToFit, I like this view better than cropped**
> * [X]  Add Play button to playlist list view to make it play through the playlist from start to finish 😸 - **I added the button but there is no functionality** 
> * [ ]  I played an episode from the playlist. At the end of the episode I expected it to play the next items in the playlist, instead it played the next episode in the series, which isn't in the playlist.

**Pull request closed without merging**